File: visitor_pattern/C/src/heap_sort.cpp
#include <inc/heap_sort.h>
#include <algorithm>

template<typename T, typename C>
void heap_sort<T,C>::sift_down(size_t idx, size_t size)
{
    C c;
    size_t l = idx * 2 + 1, r = l + 1;
    if (l >= size)
        return;
    if (r < size)
        l = c(a[l], a[r])?r:l;
    if (!c(a[idx], a[l]))
        return;
    std::swap(a[idx], a[l]);
    sift_down(l, size);
}
template<typename T, typename C>
void heap_sort<T,C>::heapify(void)
{
    for (size_t i = sz / 2; i != -1; --i)
    {
        sift_down(i, sz);
    }
}
template<typename T, typename C>
void heap_sort<T,C>::do_work(void)
{ 
    heapify();
    for (int i = sz -1; i != 0; --i)
    {
        std::swap(a[0], a[i]);
        sift_down(0, i);
    }
}

File: visitor_pattern/java/HeapSort.java
//package com.charlesq.java.visitorpattern;
import java.lang.Comparable;

/* a concrete Element for visitor pattern */

public class HeapSort<T extends Comparable<T>> implements Visitable<T>
{
    private transient T a[];
    @Override
    public void accept(Visitor<T> v)
    {
        v.visit(this);
    } 
    public void init(T []a)
    {
        this.a = a;
    }
  
    public void run()
    {
        /* produce a max heap */
        heapify(); 
        for (int i = a.length - 1; i != 0; --i)
        {
           swap(i, 0); 
           siftDown(0, i);
        }
    }
    /* recursive sift down procedure in heap sorting */
    private void siftDown(int i, int len)
    {
        int l = 2 * i + 1, r = l + 1; 
        if (l >= len)
            return;
        if (r < len)
        {
            l = a[l].compareTo(a[r]) >= 0? l: r;
        }
        if (a[i].compareTo(a[l]) >= 0)
            return; 
        swap(i, l);
        siftDown(l, len);
    }
    private void heapify()
    {
        for (int i = a.length / 2; i != -1; --i)
            siftDown(i, a.length); 
    }
    @Override
    public String toString()
    {
        return "Heap Sort";
    }
     /* A common utility method for sorting algorithm */
     void swap(int i, int j)
     {
         if (i == j)
             return;
         T temp = a[i];
         a[i] = a[j];
         a[j] = temp;
     }
}
